backend/core/src/chain.rs

```diff
@@ -257,29 +255,41 @@ impl Message for GetNodeNetworkState {
     type Result = Option<Bytes>;
 }
 
+impl NodeSource {
+    pub fn init(self, nid: NodeId, chain: Addr<Chain>) -> bool {
+        match self {
+            NodeSource::Direct { conn_id, node_connector } => {
+                node_connector
+                    .try_send(Initialize {
+                        nid,
+                        conn_id,
+                        chain,
+                    })
+                    .is_err()
+            },
+            NodeSource::Shard => {
+                // TODO
+                false
+            }
+        }
+    }
+}
+
 impl Handler<AddNode> for Chain {
     type Result = ();
 
     fn handle(&mut self, msg: AddNode, ctx: &mut Self::Context) {
         let AddNode {
             node,
-            conn_id,
-            node_connector,
+            source,
         } = msg;
         log::trace!(target: "Chain::AddNode", "New node connected. Chain '{}', node count goes from {} to {}", node.chain, self.nodes.len(), self.nodes.len() + 1);
         self.increment_label_count(&node.chain);
 
         let nid = self.nodes.add(Node::new(node));
         let chain = ctx.address();
 
-        if node_connector
-            .try_send(Initialize {
-                nid,
-                conn_id,
-                chain,
-            })
-            .is_err()
-        {
+        if source.init(nid, chain) {
```

Comment:
I think this reads a bit backwards, `if init() then remove`. Consider hoisting out the `is_err()` check here and invert the if/else.

Reply:
Yeah, I just hacked this quickly, should have left a TODO of some kind. Was thinking about returning an enum so it's clear what's happening.